### PR TITLE
Cloudinary multiple upload

### PR DESCRIPTION
**Pull request by grayey**
### What this PR does (required):
- This feature enables multiple image uploads.

### Screenshots / Videos (required):
image

image

### Any information needed to test this feature (required):
- "npm install" to obtain the cloudinary sdk
- update `.env` with the cloudinary varaibles shared on slack
- launch postman
- select method: `POST` and url: `http://localhost:3001/uploads/multiple`
- goto: Body -> form-data
- enter keys:
       - userId : "anyString"
       - multiFile: select type as File; then select multiple images
 Send

### Any issues with the current functionality (optional):
- List any problems where you were not able to complete all tasks on ticket
- List any problems this PR may cause for the project or other active PRs
- Post a screenshot/video of the problem, along with detailed console outputs where applicable

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,27 +1,30 @@
...
+
+
+#TODO add .env to your gitignore
+__snapshots__
+
+# debug & other vs code settings
+.vscode
+
+.eslintcache
+sample.env
```
Comment (moffatethan) [inline .gitignore:30]: We don't want to ignore this

File: client/.eslintrc [hunk 2]
```diff
@@ -18,6 +18,7 @@
    "react/react-in-jsx-scope": "off",
    "react/prop-types": "off",
    "react-hooks/rules-of-hooks": "error",
-   "react-hooks/exhaustive-deps": "error"
+   "react-hooks/exhaustive-deps": "error",
+   "prettier/prettier": ["error", { "endOfLine": "auto" }]
```
Comment (moffatethan) [inline client/.eslintrc:22]: I believe this is causing some issues with over `git diff`, because we have people using different operating systems when PRs are pushed up there is unnecessary changes shown in the PRs. As mentioned in another one of your PRs there is a way to ensure all operating systems utilize the same line sequences.

Comment (grayey) [reply client/.eslintrc:22]: This too.

File: .gitignore (hunk 1 above)
Comment (grayey) [reply .gitignore:30]: I will take this out and commit back.

File: server/controllers/upload.js [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+ * @param {*} param0 
+ * @returns <Promise>
+ */
+const doUpload = async ({ originalname, buffer }) => {
+    return new Promise( async (resolve, reject)=>{
+        return await cursor.upload(buffer, { originalname }).then( async (fileResponse) => {
+            resolve(fileResponse)
+         }).catch(reject)
+    })
+}
```
Comment (moffatethan) [inline server/controllers/upload.js:39]: Why not just respond with the response cursor.upload(buffer, { originalName}) gives?

Comment (moffatethan) [reply server/controllers/upload.js:39]: We don't need to wrap in a Prommise if ultimately what we're doing is just grabbing the fileResponse from the upload method.

File: server/controllers/upload.js [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+
+/**
+ * This method bulk persists uploads
+ * @param {*} instances 
+ */
+const bulkSave = async (instances) => {
+    return new Promise(async (resolve, reject) => await Upload.collection.insertMany(instances, (error, result) =>
+     error && reject(error) || resolve(result)
+     )
+    );
```
Comment (moffatethan) [inline server/controllers/upload.js:76]: As mentioned above, we don't need to be wrapping promises in promises. Just get the result from the insertMany method and return that result. The insertMany will still run asynchronously.

File: server/controllers/upload.js [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+exports.uploadSingle = asyncHandler(async(req, res, next) => {
+    const { body, file } = req;
+    const { userId } = body;
+    const { originalname, mimetype } = file;
+    const data = { userId, fileName:originalname, fileType:mimetype };
+    const upload = retrieveUpload(file);
+   
+    if(!upload){
+        res.status(HTTP_CONSTANTS.BAD_REQUEST);
+        throw new Error('Invalid file type. Please upload [jpeg, png, jpg]')
```
Comment (moffatethan) [inline server/controllers/upload.js:92]: jpeg and jpg are actually the same thing, the only difference is one is spelt longer than the other.

File: server/controllers/upload.js [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+        const { secure_url } = fileResponse;
+        const isDuplicate = await uploadExists(secure_url);
+        upload.fileUrl = secure_url;
+        await saveUpload(upload, { ...data, isDuplicate });
+        res.status(HTTP_CONSTANTS.OK).json({ upload });
+    }).catch((error)=>{
+        const { http_code, message } = error;
+        res.status(http_code || HTTP_CONSTANTS.BAD_REQUEST);
+        throw new Error(message);
+    });
```
Comment (moffatethan) [inline server/controllers/upload.js:104]: With my suggested changes above, this can be cleaned up a lot

File: server/controllers/upload.js [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+                const isDuplicate = await uploadExists(secure_url);
+                const appendUpload = () => {
+                    upload.fileUrl = secure_url;
+                    uploads.push(setProperties(upload, {...data, isDuplicate }))
+                }
+                (upload && !isDuplicate) ? appendUpload() : duplicates.push(isDuplicate);
+            }).catch((error)=>{
+                const { http_code, message } = error;
+                errors.push({ http_code, message });
+            });
```
Comment (moffatethan) [inline server/controllers/upload.js:135]: After suggested changes, this can be cleaned up a lot. You don't necessarily have to even use `.then` here. Await will pause the execution until the promise is resolved.

File: server/routes/upload.js [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+ *       required: true 
+ *       type: string 
+ *     responses:  
+ *       201: 
+ *         description: Created  
+ *       400: 
+ *         description: Bad Request  
+ *   
+ */
+router.route("/single").post([fileUploader.single('singleFile')],uploadSingle);
```
Comment (moffatethan) [inline server/routes/upload.js:34]: Middleware here doesn't need to be wrapped in an array

File: server/routes/upload.js [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+ *       required: true 
+ *       type: string 
+ *     responses:  
+ *       201: 
+ *         description: Created  
+ *       400: 
+ *         description: Bad Request  
+ *   
+ */
+router.route("/multiple").post([fileUploader.array('multiFile')],uploadMultiple);
```
Comment (moffatethan) [inline server/routes/upload.js:59]: Same here

**Review by moffatethan — CHANGES_REQUESTED** [review]
Just a few things I would like you to look at. Thank you!

File: server/routes/upload.js (hunk 8 above)
Comment (grayey) [reply server/routes/upload.js:34]: Noted. But I used an array here because I might add more middlewares in time. Like `protect`, for instance

File: server/routes/upload.js (hunk 9 above)
Comment (grayey) [reply server/routes/upload.js:59]: Same as above.

File: server/controllers/upload.js (hunk 7 above)
Comment (grayey) [reply server/controllers/upload.js:135]: Noted.

File: server/controllers/upload.js (hunk 6 above)
Comment (grayey) [reply server/controllers/upload.js:104]: Noted.

File: server/controllers/upload.js (hunk 4 above)
Comment (grayey) [reply server/controllers/upload.js:76]: Noted.

File: server/controllers/upload.js (hunk 3 above)
Comment (grayey) [reply server/controllers/upload.js:39]: Noted.

File: server/controllers/upload.js (hunk 5 above)
Comment (grayey) [reply server/controllers/upload.js:92]: Noted.

File: server/routes/upload.js (hunk 8 above)
Comment (moffatethan) [reply server/routes/upload.js:34]: Do we need to add protect on that route? Or can the parent router handle that middelware?

Comment (grayey) [reply server/routes/upload.js:34]: ah I see. Ok, noted.

File: server/controllers/upload.js (hunk 7 above)
Comment (moffatethan) [reply server/controllers/upload.js:135]: The promise wrap is not necessary as upload is already a promise. If you must catch any errors thrown by the upload promise you can use try/catch which is cleaner.
